### PR TITLE
UART IRQ examples

### PR DESCRIPTION
**Pull request by thejpster**
Demonstrates some methods for getting your HAL peripherals somewhere an interrupt routine can use them.

### CODE REVIEW

**Comment by jannic** [comment]
Very nice examples. They show the complexities involved in setting up an interrupt and passing data to and from interrupt context, without hiding it behind some opaque abstraction. At the same time, they are easy enough to just read the code and understand what's going on. I like it.
(But note that I don't have enough experience with interrupt code to judge if it's a good approach or if the code could be even simpler.)

File: rp2040-hal/examples/uart_irq_echo.rs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+/// This how we transfer the UART into the Interrupt Handler
+static GLOBAL_UART: Mutex<RefCell<Option<Uart>>> = Mutex::new(RefCell::new(None));
+
+/// Entry point to our bare-metal application.
+///
+/// The `#[entry]` macro ensures the Cortex-M start-up code calls this function
+/// as soon as all global variables are initialised.
+///
+/// The function configures the RP2040 peripherals, then writes to the UART in
+/// an inifinite loop.
```
Comment (9names) [inline rp2040-hal/examples/uart_irq_echo.rs:75]:
```suggestion
/// an infinite loop.
```

File: rp2040-hal/examples/uart_irq_tx_buffer.rs [hunk 2]
```diff
@@ -0,0 +1,285 @@
+//! # UART IRQ TX BUffer Example
```
Comment (9names) [inline rp2040-hal/examples/uart_irq_tx_buffer.rs:1]:
```suggestion
//! # UART IRQ TX Buffer Example
```

File: rp2040-hal/examples/uart_irq_tx_buffer.rs [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+type Uart = hal::uart::UartPeripheral<hal::uart::Enabled, pac::UART0>;
+
+/// This describes the queue we use for outbound UART data
+struct UartQueue {
+    mutex_cell_queue: Mutex<RefCell<Queue<u8, 64>>>,
+    interrupt: pac::Interrupt,
+}
+
+/// The linker will place this boot block at the start of our program image. We
+// need this to help the ROM bootloader get our code up and running.
```
Comment (9names) [inline rp2040-hal/examples/uart_irq_tx_buffer.rs:66]:
```suggestion
/// The linker will place this boot block at the start of our program image. We
/// need this to help the ROM bootloader get our code up and running.
```

File: rp2040-hal/examples/uart_irq_tx_buffer.rs [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    interrupt: hal::pac::Interrupt::UART0_IRQ,
+};
+
+/// Entry point to our bare-metal application.
+///
+/// The `#[entry]` macro ensures the Cortex-M start-up code calls this function
+/// as soon as all global variables are initialised.
+///
+/// The function configures the RP2040 peripherals, then writes to the UART in
+/// an inifinite loop.
```
Comment (9names) [inline rp2040-hal/examples/uart_irq_tx_buffer.rs:91]:
```suggestion
/// This function configures the RP2040 peripherals, then writes to the UART in
/// an infinite loop.
```

File: rp2040-hal/examples/uart_irq_tx_buffer.rs [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    let pins = hal::gpio::Pins::new(
+        pac.IO_BANK0,
+        pac.PADS_BANK0,
+        sio.gpio_bank0,
+        &mut pac.RESETS,
+    );
+
+    // UART TX (characters sent from RP2040) on pin 1 (GPIO0)
+    let _tx_pin = pins.gpio0.into_mode::<hal::gpio::FunctionUart>();
+    // UART RX (characters reveived by RP2040) on pin 2 (GPIO1)
```
Comment (9names) [inline rp2040-hal/examples/uart_irq_tx_buffer.rs:130]:
```suggestion
    // UART RX (characters received by RP2040) on pin 2 (GPIO1)
```

File: rp2040-hal/examples/uart_irq_tx_buffer.rs [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    cortex_m::interrupt::free(|cs| {
+        GLOBAL_UART.borrow(cs).replace(Some(uart));
+    });
+
+    // But we can blink an LED.
+    let mut led_pin = pins.gpio25.into_push_pull_output();
+
+    loop {
+        // Light the LED whilst the main thread is in the transmit routine. It
+        // shouldn't be on very long, but it will be on whilst we get enough
```
Comment (9names) [inline rp2040-hal/examples/uart_irq_tx_buffer.rs:155]:
```suggestion
        // shouldn't be on very long, but it will be on until we get enough
```

**Review by 9names — COMMENTED** [review]
Code + documentation was good, left a few suggestions for fixing typos I found while reading through.
I assume we're going to wait until #210 lands before merging this?

**Comment by thejpster** [comment]
Makes sense

**Review by 9names — APPROVED** [review]

**Pull request merged**